# **British Airways Reviews Sentiment Analysis using Machine Learning and NLP**
---
# **Step 1:Environment Setup and Dataset Loading**
### Purpose
The purpose of this step is to establish a stable and reliable development environment and to load the dataset that will be used throughout the project.

Before applying any Machine Learning or Natural Language Processing techniques, it is essential to:
- Ensure all required libraries are available
- Load the dataset correctly
- Verify the structure, size, and basic integrity of the data

In applied machine learning, **data understanding must always precede preprocessing and modelling**.

---

### Conceptual Background
Machine learning systems are fundamentally data-driven. Any errors, inconsistencies, or misunderstandings at the data-loading stage can propagate through the entire pipeline and invalidate results.

By inspecting the dataset early, we can:
- Identify which columns contain textual data suitable for NLP
- Identify potential target variables for classification
- Detect missing values or structural issues
- Ensure alignment between the dataset and the problem definition

This step supports:
- Reproducibility
- Debugging
- Ethical and responsible ML development


## ***Step 1.1: Import core libraries for data handling and visualisation***

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings as wr
wr.filterwarnings('ignore')
print("Libraries imported successfully.")

Libraries imported successfully.


#### **Expected Outcome**
- All libraries should import without errors.
- A confirmation message should be displayed indicating successful imports.

If any import fails, the environment must be fixed before proceeding.


## ***Step 1.2: Mount Google Drive and Load the dataset***

In [2]:
from google.colab import drive
drive.mount("/content/drive", force_remount=True)

Mounted at /content/drive


In [4]:
try: # Error handling for loading the dataset.
  df=pd.read_csv('/content/drive/MyDrive/Dataset/BA_AirlineReviews.csv')
  print("Dataset loaded successfully.")
except FileNotFoundError:
  print("Dataset not found. using runtime dataset")
  df =pd.read_csv('BA_AirlineReviews.csv')
except Exception as e:
  print(f"An error occurred: {e}")

Dataset not found. using runtime dataset


### ***Step 1.2.1: Exploring the Dataset***

In [5]:
df.head()

,Unnamed: 0,OverallRating,ReviewHeader,Name,Datetime,VerifiedReview,ReviewBody,TypeOfTraveller,SeatType,Route,DateFlown,SeatComfort,CabinStaffService,GroundService,ValueForMoney,Recommended,Aircraft,Food&Beverages,InflightEntertainment,Wifi&Connectivity
0,0,1.0,"""Service level far worse then Ryanair""",L Keele,19th November 2023,True,4 Hours before takeoff we received a Mail stat...,Couple Leisure,Economy Class,London to Stuttgart,November 2023,1.0,1.0,1.0,1.0,no,NaN,NaN,NaN,NaN
1,1,3.0,"""do not upgrade members based on status""",Austin Jones,19th November 2023,True,I recently had a delay on British Airways from...,Business,Economy Class,Brussels to London,November 2023,2.0,3.0,1.0,2.0,no,A320,1.0,2.0,2.0
2,2,8.0,"""Flight was smooth and quick""",M A Collie,16th November 2023,False,"Boarded on time, but it took ages to get to th...",Couple Leisure,Business Class,London Heathrow to Dublin,November 2023,3.0,3.0,4.0,3.0,yes,A320,4.0,NaN,NaN
3,3,1.0,"""Absolutely hopeless airline""",Nigel Dean,16th November 2023,True,"5 days before the flight, we were advised by B...",Couple Leisure,Economy Class,London to Dublin,December 2022,3.0,3.0,1.0,1.0,no,NaN,NaN,NaN,NaN
4,4,1.0,"""Customer Service is non existent""",Gaylynne Simpson,14th November 2023,False,"We traveled to Lisbon for our dream vacation, ...",Couple Leisure,Economy Class,London to Lisbon,November 2023,1.0,1.0,1.0,1.0,no,NaN,1.0,1.0,1.0


#### **Explanation**
Viewing the first few rows allows us to:
- Confirm that text-based fields (such as review content) are present
- Identify potential target variables (e.g., ratings or recommendation labels)
- Ensure that the data is human-readable and correctly parsed

At this stage, **no assumptions are made** about which columns will be used.
We are only observing the data.


In [6]:
# Let's look into the shape of the dataset
print("""This dataset consists of
 following total number of rows and columns\n""",df.shape)

This dataset consists of
 following total number of rows and columns
 (3701, 20)


#### **Output Analysis**
- The dataset was load without errors.
- The printed shape should indicate approximately:
  - 3,701 rows (customer reviews)
  - 20 columns (attributes describing each review)

This confirms that the dataset matches the description provided in the submitted project report.

In [8]:
df.tail()

,Unnamed: 0,OverallRating,ReviewHeader,Name,Datetime,VerifiedReview,ReviewBody,TypeOfTraveller,SeatType,Route,DateFlown,SeatComfort,CabinStaffService,GroundService,ValueForMoney,Recommended,Aircraft,Food&Beverages,InflightEntertainment,Wifi&Connectivity
3696,3696,1.0,British Airways customer review,J Robertson,29th August 2012,False,This was a bmi Regional operated flight on a R...,NaN,Economy Class,NaN,NaN,3.0,1.0,NaN,3.0,no,NaN,2.0,NaN,NaN
3697,3697,9.0,British Airways customer review,Nick Berry,28th August 2012,False,LHR to HAM. Purser addresses all club passenge...,NaN,Business Class,NaN,NaN,4.0,5.0,NaN,3.0,yes,NaN,4.0,NaN,NaN
3698,3698,5.0,British Airways customer review,Avril Barclay,12th October 2011,False,My son who had worked for British Airways urge...,NaN,Economy Class,NaN,NaN,NaN,NaN,NaN,4.0,yes,NaN,NaN,NaN,NaN
3699,3699,4.0,British Airways customer review,C Volz,11th October 2011,False,London City-New York JFK via Shannon on A318 b...,NaN,Premium Economy,NaN,NaN,1.0,3.0,NaN,1.0,no,NaN,5.0,NaN,NaN
3700,3700,4.0,British Airways customer review,Russ King,9th October 2011,False,SIN-LHR BA12 B747-436 First Class. Old aircraf...,NaN,First Class,NaN,NaN,4.0,5.0,NaN,2.0,no,NaN,5.0,2.0,NaN


#### **Explanation**
Viewing the last few rows allows us to:
- Confirm that text-based fields (such as review content) are present
- Identify potential target variables (e.g., ratings or recommendation labels)
- Ensure that the data is human-readable and correctly parsed

At this stage, **no assumptions are made** about which columns will be used.
We are only observing the data.


# **Step  2: Data Cleaning**

## ***Step 2.1: Inspect dataset structure and missing values***

In [10]:
"""The `info()` method provides critical structural information, including:
- Column names and data types
- Number of non-null values per column
- Early indication of missing data"""
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3701 entries, 0 to 3700
Data columns (total 20 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Unnamed: 0             3701 non-null   int64  
 1   OverallRating          3696 non-null   float64
 2   ReviewHeader           3701 non-null   object 
 3   Name                   3701 non-null   object 
 4   Datetime               3701 non-null   object 
 5   VerifiedReview         3701 non-null   bool   
 6   ReviewBody             3701 non-null   object 
 7   TypeOfTraveller        2930 non-null   object 
 8   SeatType               3699 non-null   object 
 9   Route                  2926 non-null   object 
 10  DateFlown              2923 non-null   object 
 11  SeatComfort            3585 non-null   float64
 12  CabinStaffService      3574 non-null   float64
 13  GroundService          2855 non-null   float64
 14  ValueForMoney          3700 non-null   float64
 15  Reco

In [12]:
df.isnull().sum() # Generalise missing values


,0
Unnamed: 0,0
OverallRating,5
ReviewHeader,0
Name,0
Datetime,0
VerifiedReview,0
ReviewBody,0
TypeOfTraveller,771
SeatType,2
Route,775


#### **Dataset Structure Summary**

The dataset contains **3,701 customer reviews** with **20 columns**.

- Textual columns such as **`ReviewBody`** and **`ReviewHeader`** are complete and suitable for NLP tasks.
- The **target-related columns** (`OverallRating` and `Recommended`) have very few or no missing values.
- Several service-related and metadata columns contain missing values, indicating they are optional for the core text classification task.

The dataset contains **3,701 rows** in total.  
The **Non-Null Count** column shows how many values are present in each column.

- If a column has **3,701 non-null values**, it contains **no missing data**
- If a column has **fewer than 3,701 non-null values**, the remaining entries are **missing values**

From the output:
- `ReviewBody` and `ReviewHeader` have **no missing values** and are reliable for NLP
- `OverallRating` has **5 missing values**
- `TypeOfTraveller`, `Route`, and `DateFlown` have **several missing values**
- `Wifi&Connectivity` has **a large number of missing values**

This confirms that the **textual review data is complete**, while some optional metadata fields are incomplete and must be handled carefully if used later.

# **Step 3: Text Column Selection for NLP**

Although the dataset contains multiple text-related columns, not all of them carry the same semantic value.

From inspection of the dataset:
- `ReviewHeader` contains short summary titles
- `ReviewBody` contains detailed, free-form customer feedback

For sentiment analysis and customer feedback classification:
- **Longer, descriptive text provides richer linguistic signals**
- Models benefit from full context, opinions, and explanations

Therefore, this project uses **`ReviewBody` as the primary text input**.


In [15]:
df['ReviewBody'] ## checking reviews

,ReviewBody
0,4 Hours before takeoff we received a Mail stat...
1,I recently had a delay on British Airways from...
2,"Boarded on time, but it took ages to get to th..."
3,"5 days before the flight, we were advised by B..."
4,"We traveled to Lisbon for our dream vacation, ..."
...,...
3696,This was a bmi Regional operated flight on a R...
3697,LHR to HAM. Purser addresses all club passenge...
3698,My son who had worked for British Airways urge...
3699,London City-New York JFK via Shannon on A318 b...


In [17]:
df['ReviewHeader'] ## checking review titles

,ReviewHeader
0,"""Service level far worse then Ryanair"""
1,"""do not upgrade members based on status"""
2,"""Flight was smooth and quick"""
3,"""Absolutely hopeless airline"""
4,"""Customer Service is non existent"""
...,...
3696,British Airways customer review
3697,British Airways customer review
3698,British Airways customer review
3699,British Airways customer review


## ***Step 3.1: Select the text column***

In [18]:
# Store review text as a separate variable
texts = df["ReviewBody"]

In [19]:
# Quick checking
texts.head()

,ReviewBody
0,4 Hours before takeoff we received a Mail stat...
1,I recently had a delay on British Airways from...
2,"Boarded on time, but it took ages to get to th..."
3,"5 days before the flight, we were advised by B..."
4,"We traveled to Lisbon for our dream vacation, ..."


#### **Explanation**
Here, we extract the `ReviewBody` column and store it in a separate variable called `texts`.

This separation:
- Makes the notebook easier to read
- Clearly distinguishes input features from target labels
- Simplifies later preprocessing and modelling steps

At this stage:
- No text cleaning is applied
- No transformations are performed
- We only confirm that the raw text is accessible


## ***Step 3.2: Inspect Raw Text Examples***

Before applying any preprocessing, it is important to **look at raw text samples**.

This helps us:
- Understand writing style and structure
- Identify noise such as punctuation, casing, or informal language
- Avoid over-cleaning that could remove useful information


In [21]:
# Displaying a few raw review examples
for i in range(4):
    print(f"Review {i+1}:")
    print(texts.iloc[i])
    print("-" * 1000)

Review 1:
4 Hours before takeoff we received a Mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time. So did the capacity of the Heathrow Airport really hit British Airways by surprise, 4h before departure? Anyhow - we took the one hour delay so what - but then we have been forced to check in our Hand luggage. I travel only with hand luggage to avoid waiting for the ultra slow processing of the checked in luggage. Overall 2h later at home than planed, with really no reason, just due to incompetent people. Service level far worse then Ryanair and triple the price. Really never again. Thanks for nothing.
-----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

#### **Interpretation of Raw Review Examples**
The printed reviews show **real, unprocessed customer feedback** written in natural language.

From these examples, we can observe that:
- Reviews vary significantly in **length**, ranging from short summaries to very long detailed complaints
- Customers use **informal language**, punctuation, and expressive wording
- Reviews often contain **multiple sentiments** within the same text (e.g. delays, staff behaviour, baggage issues)
- No obvious formatting issues (such as empty text or corrupted characters) are present

These observations confirm that:
- The text data is rich and expressive
- Context and word relationships are important for understanding sentiment
- Both classical NLP models and transformer-based models are suitable for this task

At this stage, the text is intentionally left **unprocessed** so that preprocessing decisions can be made based on observed characteristics rather than assumptions.


## ***Step 3.3: Text-Specific EDA – Review Length Analysis***

##### **Purpose**
The purpose of this step is to understand the **length characteristics** of the customer reviews.

In NLP tasks, review length matters because:
- Very short reviews may lack sufficient context
- Very long reviews may dominate learning or require truncation
- Length distribution helps guide preprocessing and model choice

This analysis is performed **before any text cleaning** so that preprocessing decisions are data-driven.


# Step 3.3.1: Compute review length (number of words)
# Calculate word count for each review


In [24]:
'''This line creates a new column called `review_word_count` by counting the number of words in each review.
The `apply()` function processes each review individually,
 `split()` separates the text into words, and
 `len()` counts how many words are present.
'''
df["review_word_count"] = df["ReviewBody"].apply(lambda x: len(x.split()))

In [25]:
# Display basic statistics
df["review_word_count"].describe()

,review_word_count
count,3701.000000
mean,160.173467
std,105.732171
min,15.000000
25%,86.000000
50%,133.000000
75%,203.000000
max,654.000000


#### **Interpretation of Review Length Statistics**

The descriptive statistics of `review_word_count` summarise how long the customer reviews are in terms of word count.

- The dataset contains **3,701 reviews**, confirming that every review has text.
- The **average review length** is approximately **160 words**, indicating that most customers provide detailed feedback.
- The **median review length** is **133 words**, showing that half of the reviews are shorter than this and half are longer.
- Most reviews fall between **86 words (25th percentile)** and **203 words (75th percentile)**.
- The **shortest review** contains **15 words**, while the **longest review** contains **654 words**, indicating some very detailed complaints or explanations.

Overall, this distribution shows that:
- Reviews are generally long enough to provide meaningful context
- There is noticeable variation in review length
- A small number of very long reviews exist, which may require truncation when using transformer-based models


## ***Step 3.4: Text-Specific EDA – Basic Word Frequency Exploration***

#### **Purpose**
The purpose of this step is to explore **which words appear most frequently** in the customer reviews *before any preprocessing is applied*.

Word frequency exploration helps us:
- Understand the dominant vocabulary used by customers
- Identify common but uninformative words (e.g. stopwords)
- Detect domain-specific terms (e.g. flight, delay, staff)
- Justify preprocessing steps such as stopword removal or lemmatisation


In [27]:
# Step 3.4.1: Tokenise text into words (simple whitespace split)

from collections import Counter

# Combine all reviews into one large string
all_text = " ".join(df["ReviewBody"])

In [28]:
# Split into individual words
all_words = all_text.split()

In [29]:
# Count word frequencies
word_counts = Counter(all_words)

In [30]:
# Display the 20 most common words
total_count=word_counts.most_common(20)
print(total_count)

[('the', 24941), ('to', 19580), ('and', 18094), ('a', 13073), ('was', 12883), ('I', 10655), ('of', 8462), ('in', 7899), ('on', 7252), ('for', 6098), ('with', 5770), ('The', 4777), ('flight', 4735), ('is', 4621), ('were', 4454), ('not', 4115), ('that', 4083), ('BA', 3955), ('but', 3767), ('at', 3656)]


### ***Step 3.4.1: Convert word frequencies to a DataFrame for readability***

In [31]:
word_freq_df = pd.DataFrame(
    total_count,
    columns=["Word", "Frequency"]
)
word_freq_df

,Word,Frequency
0,the,24941
1,to,19580
2,and,18094
3,a,13073
4,was,12883
5,I,10655
6,of,8462
7,in,7899
8,on,7252
9,for,6098


#### **Interpretation of Word Frequency Output**

The word frequency results show that the most common words in the reviews are:

- High-frequency **function words** such as *the*, *to*, *and*, *a*, *was*
- Variations of the same word due to casing (e.g. *the* and *The*)
- **Domain-specific words** such as *flight*, *BA*, *staff*
- Contrast and sentiment-related words such as *not* and *but*

This indicates that:
- Many of the most frequent words do **not carry sentiment or meaning** on their own
- Lowercasing will help merge duplicate word forms (e.g. *The* and *the*)
- Stopword removal will reduce noise in the text
- Domain-specific words should be preserved, as they are relevant to airline feedback

These observations justify the need for systematic text preprocessing in the next step.

## **Step 4: Text Preprocessing – Normalisation and Cleaning**

### ***Step 4.1: Lowercasing Text***

#### **Purpose**
Lowercasing converts all characters in the text to lowercase.

This step helps:
- Reduce vocabulary size
- Merge duplicate word forms (e.g. *The* and *the*)
- Improve consistency across the dataset

Lowercasing is a standard first step in NLP preprocessing and does not remove semantic meaning.


In [33]:
# Step 4.1: Convert all review text to lowercase

df["ReviewBody_lower"] = df["ReviewBody"].str.lower() #.str.lower() for converting every word in the column in lowercases.

In [34]:
# Inspect original vs lowercased text
for i in range (3):
  print("Original Text:")
  print(df['ReviewBody'].iloc[i])
  print("Lowercased Text:")
  print(df['ReviewBody_lower'].iloc[i])

Original Text:
4 Hours before takeoff we received a Mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time. So did the capacity of the Heathrow Airport really hit British Airways by surprise, 4h before departure? Anyhow - we took the one hour delay so what - but then we have been forced to check in our Hand luggage. I travel only with hand luggage to avoid waiting for the ultra slow processing of the checked in luggage. Overall 2h later at home than planed, with really no reason, just due to incompetent people. Service level far worse then Ryanair and triple the price. Really never again. Thanks for nothing.
Lowercased Text:
4 hours before takeoff we received a mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time. so did the capacity of the heathrow airport really hit british airways by surprise, 4h before departure? a

### **Explanation**
The original review text is preserved, and a new column `ReviewBody_lower` is created.

Observations:
- All characters are converted to lowercase
- Word content remains unchanged
- Differences caused by capitalisation are removed

This prepares the text for further preprocessing steps such as tokenisation and stopword removal.

In [35]:
df.head() #inspecting the addition of 2 new columns

,Unnamed: 0,OverallRating,ReviewHeader,Name,Datetime,VerifiedReview,ReviewBody,TypeOfTraveller,SeatType,Route,...,CabinStaffService,GroundService,ValueForMoney,Recommended,Aircraft,Food&Beverages,InflightEntertainment,Wifi&Connectivity,review_word_count,ReviewBody_lower
0,0,1.0,"""Service level far worse then Ryanair""",L Keele,19th November 2023,True,4 Hours before takeoff we received a Mail stat...,Couple Leisure,Economy Class,London to Stuttgart,...,1.0,1.0,1.0,no,NaN,NaN,NaN,NaN,125,4 hours before takeoff we received a mail stat...
1,1,3.0,"""do not upgrade members based on status""",Austin Jones,19th November 2023,True,I recently had a delay on British Airways from...,Business,Economy Class,Brussels to London,...,3.0,1.0,2.0,no,A320,1.0,2.0,2.0,206,i recently had a delay on british airways from...
2,2,8.0,"""Flight was smooth and quick""",M A Collie,16th November 2023,False,"Boarded on time, but it took ages to get to th...",Couple Leisure,Business Class,London Heathrow to Dublin,...,3.0,4.0,3.0,yes,A320,4.0,NaN,NaN,54,"boarded on time, but it took ages to get to th..."
3,3,1.0,"""Absolutely hopeless airline""",Nigel Dean,16th November 2023,True,"5 days before the flight, we were advised by B...",Couple Leisure,Economy Class,London to Dublin,...,3.0,1.0,1.0,no,NaN,NaN,NaN,NaN,255,"5 days before the flight, we were advised by b..."
4,4,1.0,"""Customer Service is non existent""",Gaylynne Simpson,14th November 2023,False,"We traveled to Lisbon for our dream vacation, ...",Couple Leisure,Economy Class,London to Lisbon,...,1.0,1.0,1.0,no,NaN,1.0,1.0,1.0,172,"we traveled to lisbon for our dream vacation, ..."


## ***Step 4.2: Removing Punctuation and Non-Alphabetic Characters***

### **Purpose**
The purpose of this step is to remove **punctuation marks and non-alphabetic characters** from the text.

Customer reviews often contain:
- Commas, full stops, and exclamation marks
- Numbers, symbols, or formatting artefacts

While these characters are useful for humans, they usually:
- Do not contribute meaningful information for traditional NLP models
- Increase vocabulary size unnecessarily

Removing them simplifies the text while preserving the core word content.

### ***Step 4.2.1: Remove punctuation and non-alphabetic characters***


In [38]:
df["ReviewBody_lower"]

,ReviewBody_lower
0,4 hours before takeoff we received a mail stat...
1,i recently had a delay on british airways from...
2,"boarded on time, but it took ages to get to th..."
3,"5 days before the flight, we were advised by b..."
4,"we traveled to lisbon for our dream vacation, ..."
...,...
3696,this was a bmi regional operated flight on a r...
3697,lhr to ham. purser addresses all club passenge...
3698,my son who had worked for british airways urge...
3699,london city-new york jfk via shannon on a318 b...


In [40]:
import re #Support for regular expressions (RE).
#This module provides regular expression matching operations similar to
#those found in Perl.  It supports both 8-bit and Unicode strings; both
#the pattern and the strings being processed can contain null bytes and
#characters outside the US ASCII range.
df['ReviewBody_clean']=df["ReviewBody_lower"].apply(lambda x: re.sub(r"[^a-z\s]", "", x)) #remove any character that is NOT a lowercase letter and NOT a space.

In [41]:
for i in range (2):
  print("Original Text:")
  print(df['ReviewBody_lower'].iloc[i])
  print("Cleaned Text:")
  print(df['ReviewBody_clean'].iloc[i])
  print("-"*1000)

Original Text:
4 hours before takeoff we received a mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time. so did the capacity of the heathrow airport really hit british airways by surprise, 4h before departure? anyhow - we took the one hour delay so what - but then we have been forced to check in our hand luggage. i travel only with hand luggage to avoid waiting for the ultra slow processing of the checked in luggage. overall 2h later at home than planed, with really no reason, just due to incompetent people. service level far worse then ryanair and triple the price. really never again. thanks for nothing.
Cleaned Text:
 hours before takeoff we received a mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time so did the capacity of the heathrow airport really hit british airways by surprise h before departure anyhow  w

#### **Explanation**
This step uses a **regular expression (regex)** to clean the text:

- `[^a-z\s]` matches any character that is:
  - NOT a lowercase letter (`a–z`)
  - NOT a whitespace character
- These characters are replaced with an empty string

As a result:
- Punctuation is removed
- Numbers and symbols are removed
- Words remain separated by spaces

## ***Step 4.3: Tokenisation and Stopword Removal***

#### **Purpose**
The purpose of this step is to convert cleaned text into **individual tokens (words)** and remove **stopwords**.

Stopwords are common words (e.g. *the*, *and*, *to*) that:
- Appear very frequently
- Carry little semantic or sentiment value on their own
- Can dominate feature representations if not removed

Tokenisation and stopword removal help models focus on **meaningful words**.

This step is especially useful for:
- Bag-of-Words
- TF-IDF
- Naïve Bayes
- SVM

Transformer models may handle stopwords internally, but we apply this step here for consistency and comparison.


In [44]:
# Step 4.3.1: Load stopwords

import nltk
"""The Natural Language Toolkit (NLTK) is an open source Python library
for Natural Language Processing.  A free online book is available.
(If you use the library for academic research, please cite the book.)

Steven Bird, Ewan Klein, and Edward Loper (2009)"""
from nltk.corpus import stopwords
"""NLTK corpus readers.  The modules in this package provide functions
that can be used to read corpus files in a variety of formats.  These
functions can be used to read both the corpus files that are
distributed in the NLTK corpus package, and corpus files that are part
of external corpora."""

# Download stopwords if not already present
nltk.download("stopwords")

stop_words = set(stopwords.words("english"))

# Inspect a few stopwords
list(stop_words)[:10]

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


['a', 'd', 'into', 'll', 'their', 'mustn', "they're", 'only', "couldn't", 'of']

#### **Explanation**
We use the **NLTK stopwords list**, which contains commonly used English words that generally do not contribute to sentiment or meaning.

The stopwords are stored in a set for:
- Faster lookup
- Efficient filtering during tokenisation


In [46]:
# Step 4.3.2: Tokenise text and remove stopwords

df["ReviewBody_tokens"] = df["ReviewBody_clean"].apply(
    lambda word: [word for word in word.split() if word not in stop_words]
)

# Inspect tokenised output
for i in range(2):
    print("Cleaned Text:")
    print(df["ReviewBody_clean"].iloc[i])
    print("\nTokens (stopwords removed):")
    print(df["ReviewBody_tokens"].iloc[i])
    print("-" * 80)

Cleaned Text:
 hours before takeoff we received a mail stating a cryptic message that there are disruptions to be expected as there is a limit on how many planes can leave at the same time so did the capacity of the heathrow airport really hit british airways by surprise h before departure anyhow  we took the one hour delay so what  but then we have been forced to check in our hand luggage i travel only with hand luggage to avoid waiting for the ultra slow processing of the checked in luggage overall h later at home than planed with really no reason just due to incompetent people service level far worse then ryanair and triple the price really never again thanks for nothing

Tokens (stopwords removed):
['hours', 'takeoff', 'received', 'mail', 'stating', 'cryptic', 'message', 'disruptions', 'expected', 'limit', 'many', 'planes', 'leave', 'time', 'capacity', 'heathrow', 'airport', 'really', 'hit', 'british', 'airways', 'surprise', 'h', 'departure', 'anyhow', 'took', 'one', 'hour', 'delay

**Output Analysis**

The output shows the cleaned review text alongside the resulting list of tokens after stopword removal.

From the examples, we can observe that:
- Common function words such as *the*, *and*, *to*, *was* have been removed
- Domain-specific and sentiment-related words such as *delay*, *luggage*, *staff*, *unhelpful*, *worse* are retained
- Repeated important terms (e.g. *delay*, *luggage*) remain, which is useful for learning patterns
- The text is now represented as a list of meaningful tokens rather than raw sentences

This confirms that:
- Tokenisation is functioning as expected
- Stopword removal successfully reduces noise
- The remaining tokens preserve the semantic content of customer feedback

At this stage, the text is well-prepared for numerical feature extraction in later steps.

## ***Step 4.4: Optional Lemmatisation***
**Purpose**

The purpose of this step is to decide whether **lemmatisation** should be applied to the tokenised text.

Lemmatisation reduces words to their **base (dictionary) form**, which can:
- Reduce vocabulary size
- Group different grammatical forms of the same word
- Improve generalisation for classical machine learning models
---
This step is marked *optional* because its usefulness depends on the task and model choice.
---
Lemmatisation is particularly helpful when:
- Using **Bag-of-Words** or **TF-IDF**
- Vocabulary size needs to be reduced
- Different word forms carry the same meaning

However, for **transformer models** (e.g. BERT):
- Contextual embeddings already handle word forms
- Lemmatisation may remove useful grammatical signals

---
For this project, lemmatisation can be applied **only to the classical ML pipeline**, while keeping raw or lightly cleaned text for transformers.

In [48]:
### ***Step 4.4.1: Load WordNet lemmatiser***


In [49]:

"""
We use NLTK’s **WordNetLemmatizer**, which applies dictionary-based rules to convert words into their base form.
At this stage:
- We are not yet changing the dataset
- We first test lemmatisation on a small sample
"""
from nltk.stem import WordNetLemmatizer
nltk.download("wordnet")
nltk.download("omw-1.4")

lemmatizer = WordNetLemmatizer()

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


In [50]:
# Step 4.4.2: Apply lemmatisation to token lists (classical ML only)

df["ReviewBody_lemmatized"] = df["ReviewBody_tokens"].apply(
    lambda tokens: [lemmatizer.lemmatize(word) for word in tokens]
)

# Inspect before and after
for i in range(2):
    print("Tokens:")
    print(df["ReviewBody_tokens"].iloc[i])
    print("\nLemmatized Tokens:")
    print(df["ReviewBody_lemmatized"].iloc[i])
    print("-" * 80)

Tokens:
['hours', 'takeoff', 'received', 'mail', 'stating', 'cryptic', 'message', 'disruptions', 'expected', 'limit', 'many', 'planes', 'leave', 'time', 'capacity', 'heathrow', 'airport', 'really', 'hit', 'british', 'airways', 'surprise', 'h', 'departure', 'anyhow', 'took', 'one', 'hour', 'delay', 'forced', 'check', 'hand', 'luggage', 'travel', 'hand', 'luggage', 'avoid', 'waiting', 'ultra', 'slow', 'processing', 'checked', 'luggage', 'overall', 'h', 'later', 'home', 'planed', 'really', 'reason', 'due', 'incompetent', 'people', 'service', 'level', 'far', 'worse', 'ryanair', 'triple', 'price', 'really', 'never', 'thanks', 'nothing']

Lemmatized Tokens:
['hour', 'takeoff', 'received', 'mail', 'stating', 'cryptic', 'message', 'disruption', 'expected', 'limit', 'many', 'plane', 'leave', 'time', 'capacity', 'heathrow', 'airport', 'really', 'hit', 'british', 'airway', 'surprise', 'h', 'departure', 'anyhow', 'took', 'one', 'hour', 'delay', 'forced', 'check', 'hand', 'luggage', 'travel', 'hand

# **Design Decision**

For this project:
- **Lemmatisation will be applied to classical ML models** (TF-IDF + SVM / NB / KNN)
- **Transformer models will use minimally processed text**

This allows a fair comparison between:
- Traditional feature engineering approaches
- Modern contextual embedding approaches


# **Step 5: Feature Extraction Using TF-IDF (Classical ML Pipeline)**

### Purpose
Machine learning models cannot work directly with text.  
The purpose of this step is to convert preprocessed text into **numerical feature vectors** using **TF-IDF (Term Frequency–Inverse Document Frequency)**.

TF-IDF captures:
- How important a word is **within a review**
- How unique that word is **across all reviews**

This representation is widely used in classical NLP pipelines and provides a strong baseline for text classification.
---
- **Term Frequency (TF)**  
  How often a word appears in a document

- **Inverse Document Frequency (IDF)**  
  How rare the word is across all documents

The intuition is:
- Words that appear frequently in a review but rarely across the dataset are **more informative**
- Very common words receive lower importance

TF-IDF helps models focus on **discriminative words** rather than frequent but uninformative ones.


In [53]:
# Step 5.1: Prepare text for TF-IDF

# Join lemmatized tokens back into strings (classical ML only)
df["ReviewBody_tfidf"] = df["ReviewBody_lemmatized"].apply(lambda tokens: " ".join(tokens))

# Inspect the prepared text
df["ReviewBody_tfidf"].head()

,ReviewBody_tfidf
0,hour takeoff received mail stating cryptic mes...
1,recently delay british airway bru lhr due staf...
2,boarded time took age get runway due congestio...
3,day flight advised ba cancelled asked u rebook...
4,traveled lisbon dream vacation cruise portugal...


### **Explanation**
TF-IDF vectorisers expect text in **string format**, not token lists.

In this step:
- Token lists are joined back into a single string
- Words are separated by spaces
- The text remains fully preprocessed and ready for vectorisation

This representation will be used **only for classical ML models**.


## ***Step 5.2: Create TF-IDF vectoriser***

In [56]:
from sklearn.feature_extraction.text import TfidfVectorizer

tfidf_vectorizer = TfidfVectorizer(
    max_features=2000,    # limit vocabulary size
    ngram_range=(1, 2)    # unigrams + bigrams
)

#### **Explanation**
The TF-IDF vectoriser is configured with:

- `max_features=5000`  
  Limits vocabulary size to reduce dimensionality and overfitting

- `ngram_range=(1, 2)`  
  Includes:
  - Unigrams (single words)
  - Bigrams (two-word combinations, e.g. *flight delay*)

This configuration balances expressiveness and computational efficiency.

## ***Step 5.3: Fit and transform the text data***

In [57]:
X_tfidf = tfidf_vectorizer.fit_transform(df["ReviewBody_tfidf"])

# Inspect feature matrix shape
X_tfidf.shape

(3701, 2000)

### **Interpretation of TF-IDF Feature Matrix Shape**

The output `(3701, 2000)` represents the shape of the TF-IDF feature matrix.

- **3,701 rows** correspond to the number of customer reviews in the dataset
- **2,000 columns** represent the TF-IDF features (words or n-grams) selected by the vectoriser

Each review is now represented as a fixed-length numerical vector, where feature values indicate the importance of specific terms within the review relative to the entire dataset.

This transformation enables classical machine learning models to process textual data.


## ***Step 5.5: Class Imbalance Analysis***

### **Purpose**
The purpose of this step is to examine the **distribution of sentiment classes** in the dataset.

Understanding class imbalance is critical because:
- Imbalanced classes can bias model predictions
- Accuracy alone may become misleading
- Evaluation metrics must be chosen carefully

This analysis ensures that model performance is interpreted correctly.


In [59]:
df.columns


Index(['Unnamed: 0', 'OverallRating', 'ReviewHeader', 'Name', 'Datetime',
       'VerifiedReview', 'ReviewBody', 'TypeOfTraveller', 'SeatType', 'Route',
       'DateFlown', 'SeatComfort', 'CabinStaffService', 'GroundService',
       'ValueForMoney', 'Recommended', 'Aircraft', 'Food&Beverages',
       'InflightEntertainment', 'Wifi&Connectivity', 'review_word_count',
       'ReviewBody_lower', 'ReviewBody_clean', 'ReviewBody_tokens',
       'ReviewBody_lemmatized', 'ReviewBody_tfidf'],
      dtype='object')

### ***Step 5.5.1: Create / verify sentiment labels***


In [61]:
# Create sentiment labels from OverallRating
def map_sentiment(rating): # Overall rating scale is taken as 1-10
    if rating <= 3:
        return "Negative"
    elif rating <= 6:
        return "Neutral"
    else:
        return "Positive"

In [62]:
df["Sentiment"] = df["OverallRating"].apply(map_sentiment)

In [63]:
df["Sentiment"].head()

,Sentiment
0,Negative
1,Negative
2,Positive
3,Negative
4,Negative


In [64]:
df["Sentiment"].dtype #Object datatype

dtype('O')

- **The Sentiment column was temporarily overwritten with boolean values during a missing-value check.**
- **It was recreated from the OverallRating column to restore categorical sentiment labels (Negative, Neutral, Positive) required for classification.**
---
**Key Takeaway:**
Checking missing values (notna()) should never be assigned back to a semantic column.
Always inspect, never overwrite.


In [66]:
df["Sentiment"].value_counts()

,count
Sentiment,
Negative,1695
Positive,1328
Neutral,678


#### **Interpretation of Sentiment Class Distribution:**

The class distribution shows that the dataset is not evenly balanced across sentiment categories.

- **Negative** reviews are the most frequent (1,695 samples)
- **Positive** reviews are slightly fewer (1,328 samples)
- **Neutral** reviews are the least represented (678 samples)

This indicates a **moderate class imbalance**, with Neutral sentiment being underrepresented compared to Negative and Positive classes.

Such imbalance is common in real-world customer review datasets and must be considered during model evaluation.

#### **Modelling Implications**

Due to the observed class imbalance:
- Accuracy alone will not be sufficient to evaluate model performance
- Evaluation will prioritise **precision, recall, and F1-score**
- A **stratified train–test split** will be used to preserve class proportions
- Class weighting may be considered for certain classifiers if needed

## ***Step 5.6: Removing Unnecessary Columns***
### **Purpose**
At this stage, the dataset contains many columns that are **not required** for sentiment classification.

The goal of this step is to:
- Retain only the columns needed for modelling and evaluation
- Remove metadata and unused features
- Reduce memory usage and improve clarity of the dataset

This ensures a clean and focused dataset for machine learning

In [69]:
df.columns

Index(['Unnamed: 0', 'OverallRating', 'ReviewHeader', 'Name', 'Datetime',
       'VerifiedReview', 'ReviewBody', 'TypeOfTraveller', 'SeatType', 'Route',
       'DateFlown', 'SeatComfort', 'CabinStaffService', 'GroundService',
       'ValueForMoney', 'Recommended', 'Aircraft', 'Food&Beverages',
       'InflightEntertainment', 'Wifi&Connectivity', 'review_word_count',
       'ReviewBody_lower', 'ReviewBody_clean', 'ReviewBody_tokens',
       'ReviewBody_lemmatized', 'ReviewBody_tfidf', 'Sentiment'],
      dtype='object')

## ***Step 5.6: Create a clean modelling DataFrame***

In [70]:
columns_to_keep = [
    "ReviewBody_tfidf",
    "Sentiment",
    "OverallRating"
]
df_model = df[columns_to_keep].copy()
# Verify result
df_model.head()

,ReviewBody_tfidf,Sentiment,OverallRating
0,hour takeoff received mail stating cryptic mes...,Negative,1.0
1,recently delay british airway bru lhr due staf...,Negative,3.0
2,boarded time took age get runway due congestio...,Positive,8.0
3,day flight advised ba cancelled asked u rebook...,Negative,1.0
4,traveled lisbon dream vacation cruise portugal...,Negative,1.0


In [71]:
df_model.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3701 entries, 0 to 3700
Data columns (total 3 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   ReviewBody_tfidf  3701 non-null   object 
 1   Sentiment         3701 non-null   object 
 2   OverallRating     3696 non-null   float64
dtypes: float64(1), object(2)
memory usage: 86.9+ KB


# **Step 6: Stratified Train–Test Split**

### **Purpose**
Before training machine learning models, the dataset must be split into
**training** and **testing** subsets.

A stratified split is used to:
- Preserve the original class distribution in both sets
- Prevent bias caused by class imbalance
- Ensure fair and reliable model evaluation

This step prepares the data for supervised learning.
---
From the previous analysis, the sentiment classes are not evenly distributed.
Without stratification:
- Minority classes (e.g. Neutral) may be underrepresented in the test set
- Model performance metrics may become unreliable

Stratification ensures that each sentiment class is represented proportionally
in both training and testing data.